test/integration/authorize/list_app_keys_test.rb

```diff
@@ -0,0 +1,132 @@
+    assert !app_keys.nil?
+    assert_equal @application_nokeys.id, app_keys['app']
+    assert_equal @service.id, app_keys['svc']
+
+    keys = xml_resp.search("app_keys key").map { |n| n[:id] }
+
+    assert keys.empty?
+  end
+
+  test 'calling authorize with the list_app_keys extension lists the application keys even if no app_key is specified' do
+    get '/transactions/authorize.xml',
+        { provider_key: @provider_key, app_id: @application.id },
+        'HTTP_3SCALE_OPTIONS' => Extensions::LIST_APP_KEYS
+
+    xml_resp = Nokogiri::XML(last_response.body)
+    app_keys = xml_resp.at 'app_keys'
+    assert !app_keys.nil?
+    assert_equal @application.id, app_keys['app']
+    assert_equal @service.id, app_keys['svc']
+
+    keys = xml_resp.search("app_keys key").map { |n| n[:id] }
+
+    assert_equal @keys, keys.sort
+  end
+
+  test 'calling authorize with the list_app_keys extension lists the application keys even if app_key is invalid' do
+    get '/transactions/authorize.xml',
+        { provider_key: @provider_key, app_id: @application.id,
+          app_key: 'invalid-key' },
+        'HTTP_3SCALE_OPTIONS' => Extensions::LIST_APP_KEYS
+
+    xml_resp = Nokogiri::XML(last_response.body)
+    app_keys = xml_resp.at 'app_keys'
+    assert !app_keys.nil?
+    assert_equal @application.id, app_keys['app']
+    assert_equal @service.id, app_keys['svc']
+
+    keys = xml_resp.search("app_keys key").map { |n| n[:id] }
+
+    assert_equal @keys, keys.sort
+  end
+
+  test 'calling authorize with the list_app_keys extension lists the application keys when specifying a valid app_key' do
+    get '/transactions/authorize.xml',
+        { provider_key: @provider_key, app_id: @application.id,
+          app_key: @keys.sample },
+        'HTTP_3SCALE_OPTIONS' => Extensions::LIST_APP_KEYS
+
+    xml_resp = Nokogiri::XML(last_response.body)
+    app_keys = xml_resp.at 'app_keys'
+    assert !app_keys.nil?
+    assert_equal @application.id, app_keys['app']
+    assert_equal @service.id, app_keys['svc']
+
+    keys = xml_resp.search("app_keys key").map { |n| n[:id] }
+
+    assert_equal @keys, keys.sort
+  end
+
+  test "calling authorize with the list_app_keys extension lists a maximum of #{Transactor::Status.const_get(:LIST_APP_KEYS_MAX)} application keys" do
+    max_keys = Transactor::Status.const_get :LIST_APP_KEYS_MAX
+
+    (@keys.size + 1).upto(max_keys + 1) do |i|
+      @application.create_key "key_#{i}"
+    end
+
+    get '/transactions/authorize.xml',
+        { provider_key: @provider_key, app_id: @application.id,
+          app_key: @keys.sample },
+        'HTTP_3SCALE_OPTIONS' => Extensions::LIST_APP_KEYS
+
+    xml_resp = Nokogiri::XML(last_response.body)
+    app_keys = xml_resp.at 'app_keys'
+    assert !app_keys.nil?
+    assert_equal @application.id, app_keys['app']
+    assert_equal @service.id, app_keys['svc']
+
+    keys = xml_resp.search("app_keys key").map { |n| n[:id] }
+
+    assert_equal max_keys, keys.size
+  end
+end
```

Comment:
Should we test that this does not return anything when service and app_id are valid but the provider key is not?

Reply:
I don't think so - they are unrelated circumstances that should have no connection (and the related ones should be tested already). If we went that route then we could start a explosion of test combinations all over the place that would arguably provide little value.